File: server/csharp/MetaObjects.Codegen/Runtime/FilterPredicate.cs
// FilterPredicate — substrate-agnostic record describing one parsed filter
// predicate (field, op, coerced value). Returned by FilterParser; consumed by
// dispatchers like EfCoreFilterDispatch that translate it to a substrate query.
//
// Part of the codegen runtime surface — consumers reference this assembly at
// runtime in their ASP.NET Core host. Mirrors Java codegen-spring's
// runtime/FilterPredicate.java.

namespace MetaObjects.Codegen.Runtime;

/// <summary>
/// A single parsed + validated filter predicate from the FR-009 URL grammar
/// (<c>filter[&lt;field&gt;][&lt;op&gt;]=&lt;value&gt;</c>).
/// </summary>
/// <param name="Field">Allowlisted field name (already validated).</param>
/// <param name="Op">Allowlisted operator (already validated against the field's subtype).</param>
/// <param name="Value">
/// The coerced value: <see langword="string"/> for scalar ops on string-shaped
/// subtypes, a <see cref="System.Collections.Generic.List{T}"/> of strings for
/// <c>in</c>, a <see langword="bool"/> for <c>isNull</c>. The dispatcher is
/// responsible for any further conversion to the column type.
/// </param>
public sealed record FilterPredicate(string Field, string Op, object? Value);

File: server/java/codegen-spring/src/main/java/com/metaobjects/generator/spring/runtime/FilterPredicate.java
package com.metaobjects.generator.spring.runtime;

/**
 * A single parsed + validated filter predicate produced by
 * {@link FilterParser#parse}. Carries the field name, the operator (one of
 * the 9 cross-port FR-009 ops: {@code eq, ne, gt, gte, lt, lte, in, like,
 * isNull}), and the coerced value (a {@code String} for the scalar ops, a
 * {@code List<String>} for {@code in}, a {@code Boolean} for {@code isNull}).
 *
 * <p>The repository implementation is responsible for translating this into
 * its persistence DSL — there is no built-in JPA / jOOQ binding here so the
 * runtime stays substrate-agnostic.</p>
 */
public record FilterPredicate(String field, String op, Object value) {}
